Fix #256, #278

MaceWindu opened the pull request:
This PR addresses #256 and #278 and includes following changes:

- [x] Memory leak tests for #256 
- [x] Test name deduplication in base fixture
- [x] Create performance tests with and without caching
- [x] Add cache disable option
- [x] Add cache cleanup method
- [x] Reduce cache lock latency

#256 fix:
As it is not possible to fix this issue in linq2db without complete rewrite of linq queries handling, two new options added to fix it on client side:
 - `Configuration.Linq.DisableQueryCache` configuration flag to completely disable linq query cache
 - `Query<TEntity>.ClearCache()` method to clear linq query cache for specific entity type on user request
- `LinqToDB.Linq.NoLinqCache` class to explicitly disable caching of queries in current scope

#278 fix:
 - ~~Fixed use of incorrect critical section synchronization objects for non-linq DML queries (`ObjectOperation<T>` class)~~ doesn't exist anymore
 - redesigned linq query cache and reduced lock area at some memory/cpu cost

MaceWindu commented on Tests/Linq/UserTests/Issue256Tests.cs:
```diff
@@ -192,7 +196,7 @@ private static void UpdateWhere(ITestDataContext db, byte[] value, int calls)
 			var expected = true;
 			while (calls > 0)
 			{
-				query.Set(_ => _.BoolValue, _ => !_.BoolValue || true).Update();
```
BTW, this was an error in test, but linq2db generates invalid SQL for query like that. Didn't digged into it but looks like it over-optimized expression.

MaceWindu commented:
Aww, this suck. As expected - .net core GC/runtime behaves somewhat differently. Will see if I can fix it, otherwise we can disable those tests for .net core.

MaceWindu commented:
Disabled tests for .net core as I have no idea why they do not pass:
- issue in framework
- GC behavior differences
- .net core-specific memory leak in linq2db
Maybe we should use newer .net core version on Travis? Now it is 1.0.0-preview2-003156 which is pretty old.

MaceWindu commented:
That was stupid :)
Just noticed that failing tests were executed under desktop framework, not .net core.

MaceWindu commented:
Ok, in the end I've disabled memory tests for Mono because they are wery flaky due to Mono GC behavior.
At least by adding more calls to GC.Collect() I was able to significantly decrease number of failing tests which means that issue fix also works under Mono.

ili commented on Source/Linq/Query.cs:
```diff
@@ -97,6 +97,8 @@ void ClearParameters()
 				foreach (var sqlParameter in query.Parameters)
 					sqlParameter.Expression = null;
+
+			Expression = null;
```
This totally breaks query cache.  See `Query.Compare` it always returns false for `Expression==null`
I think to solve both of this issues we'll need:
* use WeakReference
* store not Expression but something like expression fingerprint

MaceWindu replied:
I will add tests to assert that cache is not broken and will try to find better solution

ili commented:
@MaceWindu what are the roots for memory leak, only `ConstantExpression` or may be some others?

MaceWindu commented:
Right, ConstantExpression

ili commented:
We can make expression transformation, setting all constants to their default values. And store and compare this converted expression trees in query cache but this needs great attention not to make fatal cache hit, for example if we'll suppose to be the same `where a == 1` and `where a == 2` this would be...

sdanyliv commented:
Looks like it is not trivial problem. Maybe we should introduce something like Sql.NoCache
``` csharp
var data = db.Table1
    .Where(_ => db.Table2.Select(_ => _.Id).Contains(_.Id))
    .Select(_ => new SomeObject(Sql.NoCache(bigObject), _.Field1, _.Field2))
    .ToList();
```

ili commented:
Using `NoCache` is similar with
```sql
var value = bigObjet.Field;

var data = db.Big.Where(_ => _.Field == value);
```

MaceWindu commented:
I've checked code and agree that it is not something that could be solved easily. For now I want to concentrate on merge feature, so let's freeze it for now.

If anyone else suffering from this issue, we can introduce temporary solution with global flag to disable query cache.

If you want I can extract unrelated fixes (duplicate parameters and test names) as separate PRs.

MaceWindu commented:
Current testrun stats for 10 000 queries in 3 modes: with cache (cache=1), without cache(cache=0) and with cache cleanup after each query (clear=1):
`set` - type of query, `Mixed` set contains all queries, executed at random
`threads` - number of parallel threads
```
TestPerformance.set=Delete.threads=01.cache=0.clear=0: 00:00:03.1633770
TestPerformance.set=Delete.threads=01.cache=1.clear=0: 00:00:00.1476668
TestPerformance.set=Delete.threads=01.cache=1.clear=1: 00:00:03.1383135
TestPerformance.set=Delete.threads=02.cache=0.clear=0: 00:00:01.6476211
TestPerformance.set=Delete.threads=02.cache=1.clear=0: 00:00:00.0775755
TestPerformance.set=Delete.threads=02.cache=1.clear=1: 00:00:03.1208017
TestPerformance.set=Delete.threads=05.cache=0.clear=0: 00:00:01.0483625
TestPerformance.set=Delete.threads=05.cache=1.clear=0: 00:00:00.0460419
TestPerformance.set=Delete.threads=05.cache=1.clear=1: 00:00:03.1436935
TestPerformance.set=Delete.threads=10.cache=0.clear=0: 00:00:01.0786684
TestPerformance.set=Delete.threads=10.cache=1.clear=0: 00:00:00.0470447
TestPerformance.set=Delete.threads=10.cache=1.clear=1: 00:00:03.0163357
TestPerformance.set=Delete.threads=20.cache=0.clear=0: 00:00:01.2221518
TestPerformance.set=Delete.threads=20.cache=1.clear=0: 00:00:00.0430445
TestPerformance.set=Delete.threads=20.cache=1.clear=1: 00:00:03.0308492
TestPerformance.set=Insert.threads=01.cache=0.clear=0: 00:00:03.7430560
TestPerformance.set=Insert.threads=01.cache=1.clear=0: 00:00:00.1936817
TestPerformance.set=Insert.threads=01.cache=1.clear=1: 00:00:03.9444249
TestPerformance.set=Insert.threads=02.cache=0.clear=0: 00:00:02.0232525
TestPerformance.set=Insert.threads=02.cache=1.clear=0: 00:00:00.1121967
TestPerformance.set=Insert.threads=02.cache=1.clear=1: 00:00:03.8694097
TestPerformance.set=Insert.threads=05.cache=0.clear=0: 00:00:01.1515887
TestPerformance.set=Insert.threads=05.cache=1.clear=0: 00:00:00.1727689
TestPerformance.set=Insert.threads=05.cache=1.clear=1: 00:00:03.6725647
TestPerformance.set=Insert.threads=10.cache=0.clear=0: 00:00:01.1257428
TestPerformance.set=Insert.threads=10.cache=1.clear=0: 00:00:00.0770749
TestPerformance.set=Insert.threads=10.cache=1.clear=1: 00:00:03.9311884
TestPerformance.set=Insert.threads=20.cache=0.clear=0: 00:00:01.3313194
TestPerformance.set=Insert.threads=20.cache=1.clear=0: 00:00:00.0840827
TestPerformance.set=Insert.threads=20.cache=1.clear=1: 00:00:04.0433009
TestPerformance.set=InsertOrUpdate.threads=01.cache=0.clear=0: 00:00:08.4307009
TestPerformance.set=InsertOrUpdate.threads=01.cache=1.clear=0: 00:00:00.3082875
TestPerformance.set=InsertOrUpdate.threads=01.cache=1.clear=1: 00:00:08.4291711
TestPerformance.set=InsertOrUpdate.threads=02.cache=0.clear=0: 00:00:04.5937674
TestPerformance.set=InsertOrUpdate.threads=02.cache=1.clear=0: 00:00:00.1690381
TestPerformance.set=InsertOrUpdate.threads=02.cache=1.clear=1: 00:00:07.9915551
TestPerformance.set=InsertOrUpdate.threads=05.cache=0.clear=0: 00:00:02.7018405
TestPerformance.set=InsertOrUpdate.threads=05.cache=1.clear=0: 00:00:00.1226209
TestPerformance.set=InsertOrUpdate.threads=05.cache=1.clear=1: 00:00:06.9073238
TestPerformance.set=InsertOrUpdate.threads=10.cache=0.clear=0: 00:00:02.5195703
TestPerformance.set=InsertOrUpdate.threads=10.cache=1.clear=0: 00:00:00.1257943
TestPerformance.set=InsertOrUpdate.threads=10.cache=1.clear=1: 00:00:06.4690738
TestPerformance.set=InsertOrUpdate.threads=20.cache=0.clear=0: 00:00:03.0074984
TestPerformance.set=InsertOrUpdate.threads=20.cache=1.clear=0: 00:00:00.1351309
TestPerformance.set=InsertOrUpdate.threads=20.cache=1.clear=1: 00:00:06.5741832
TestPerformance.set=InsertWithIdentity.threads=01.cache=0.clear=0: 00:00:03.7432324
TestPerformance.set=InsertWithIdentity.threads=01.cache=1.clear=0: 00:00:00.1806670
TestPerformance.set=InsertWithIdentity.threads=01.cache=1.clear=1: 00:00:03.8214272
TestPerformance.set=InsertWithIdentity.threads=02.cache=0.clear=0: 00:00:02.0606629
TestPerformance.set=InsertWithIdentity.threads=02.cache=1.clear=0: 00:00:00.1056027
TestPerformance.set=InsertWithIdentity.threads=02.cache=1.clear=1: 00:00:03.7569458
TestPerformance.set=InsertWithIdentity.threads=05.cache=0.clear=0: 00:00:01.1679536
TestPerformance.set=InsertWithIdentity.threads=05.cache=1.clear=0: 00:00:00.1351318
TestPerformance.set=InsertWithIdentity.threads=05.cache=1.clear=1: 00:00:03.8177450
TestPerformance.set=InsertWithIdentity.threads=10.cache=0.clear=0: 00:00:01.1470007
TestPerformance.set=InsertWithIdentity.threads=10.cache=1.clear=0: 00:00:00.0795789
TestPerformance.set=InsertWithIdentity.threads=10.cache=1.clear=1: 00:00:03.8996650
TestPerformance.set=InsertWithIdentity.threads=20.cache=0.clear=0: 00:00:01.3843606
TestPerformance.set=InsertWithIdentity.threads=20.cache=1.clear=0: 00:00:00.0880885
TestPerformance.set=InsertWithIdentity.threads=20.cache=1.clear=1: 00:00:03.9207327
TestPerformance.set=Mixed.threads=01.cache=0.clear=0: 00:00:04.9861541
TestPerformance.set=Mixed.threads=01.cache=1.clear=0: 00:00:00.2287255
TestPerformance.set=Mixed.threads=01.cache=1.clear=1: 00:00:05.0583913
TestPerformance.set=Mixed.threads=02.cache=0.clear=0: 00:00:02.7201413
TestPerformance.set=Mixed.threads=02.cache=1.clear=0: 00:00:00.5393585
TestPerformance.set=Mixed.threads=02.cache=1.clear=1: 00:00:04.2821070
TestPerformance.set=Mixed.threads=05.cache=0.clear=0: 00:00:01.5625054
TestPerformance.set=Mixed.threads=05.cache=1.clear=0: 00:00:00.8408237
TestPerformance.set=Mixed.threads=05.cache=1.clear=1: 00:00:04.3567296
TestPerformance.set=Mixed.threads=10.cache=0.clear=0: 00:00:01.6821778
TestPerformance.set=Mixed.threads=10.cache=1.clear=0: 00:00:00.9939559
TestPerformance.set=Mixed.threads=10.cache=1.clear=1: 00:00:04.7940837
TestPerformance.set=Mixed.threads=20.cache=0.clear=0: 00:00:01.7849891
TestPerformance.set=Mixed.threads=20.cache=1.clear=0: 00:00:01.6495573
TestPerformance.set=Mixed.threads=20.cache=1.clear=1: 00:00:04.7750251
TestPerformance.set=Select.threads=01.cache=0.clear=0: 00:00:04.2200797
TestPerformance.set=Select.threads=01.cache=1.clear=0: 00:00:00.1681666
TestPerformance.set=Select.threads=01.cache=1.clear=1: 00:00:04.1361520
TestPerformance.set=Select.threads=02.cache=0.clear=0: 00:00:02.1851822
TestPerformance.set=Select.threads=02.cache=1.clear=0: 00:00:00.0986891
TestPerformance.set=Select.threads=02.cache=1.clear=1: 00:00:03.8439029
TestPerformance.set=Select.threads=05.cache=0.clear=0: 00:00:01.2125493
TestPerformance.set=Select.threads=05.cache=1.clear=0: 00:00:00.1631646
TestPerformance.set=Select.threads=05.cache=1.clear=1: 00:00:04.0130077
TestPerformance.set=Select.threads=10.cache=0.clear=0: 00:00:01.4384117
TestPerformance.set=Select.threads=10.cache=1.clear=0: 00:00:00.0470398
TestPerformance.set=Select.threads=10.cache=1.clear=1: 00:00:03.8381014
TestPerformance.set=Select.threads=20.cache=0.clear=0: 00:00:01.6996019
TestPerformance.set=Select.threads=20.cache=1.clear=0: 00:00:00.0495488
TestPerformance.set=Select.threads=20.cache=1.clear=1: 00:00:03.9602155
TestPerformance.set=Update.threads=01.cache=0.clear=0: 00:00:06.4798127
TestPerformance.set=Update.threads=01.cache=1.clear=0: 00:00:00.3288071
TestPerformance.set=Update.threads=01.cache=1.clear=1: 00:00:06.4450120
TestPerformance.set=Update.threads=02.cache=0.clear=0: 00:00:03.5386675
TestPerformance.set=Update.threads=02.cache=1.clear=0: 00:00:00.2022004
TestPerformance.set=Update.threads=02.cache=1.clear=1: 00:00:06.3029057
TestPerformance.set=Update.threads=05.cache=0.clear=0: 00:00:02.1489152
TestPerformance.set=Update.threads=05.cache=1.clear=0: 00:00:00.1111087
TestPerformance.set=Update.threads=05.cache=1.clear=1: 00:00:05.3740044
TestPerformance.set=Update.threads=10.cache=0.clear=0: 00:00:01.9361623
TestPerformance.set=Update.threads=10.cache=1.clear=0: 00:00:00.1071043
TestPerformance.set=Update.threads=10.cache=1.clear=1: 00:00:05.4130865
TestPerformance.set=Update.threads=20.cache=0.clear=0: 00:00:02.6782097
TestPerformance.set=Update.threads=20.cache=1.clear=0: 00:00:00.1071049
TestPerformance.set=Update.threads=20.cache=1.clear=1: 00:00:05.0192087

```

MaceWindu commented:
LINQ queries cache lock optimizations results:

BEFORE
```
TestPerformance.set=Delete.threads=01.cache=1.clear=0: 00:00:01.3603706
TestPerformance.set=Delete.threads=02.cache=1.clear=0: 00:00:00.7447297
TestPerformance.set=Delete.threads=05.cache=1.clear=0: 00:00:00.4104036
TestPerformance.set=Delete.threads=10.cache=1.clear=0: 00:00:00.3943918
TestPerformance.set=Delete.threads=20.cache=1.clear=0: 00:00:00.3743655
TestPerformance.set=DeleteObject.threads=01.cache=1.clear=0: 00:00:00.1286469
TestPerformance.set=DeleteObject.threads=02.cache=1.clear=0: 00:00:00.1243257
TestPerformance.set=DeleteObject.threads=05.cache=1.clear=0: 00:00:00.2112029
TestPerformance.set=DeleteObject.threads=10.cache=1.clear=0: 00:00:00.2429102
TestPerformance.set=DeleteObject.threads=20.cache=1.clear=0: 00:00:00.2482329
TestPerformance.set=Insert.threads=01.cache=1.clear=0: 00:00:02.0713722
TestPerformance.set=Insert.threads=02.cache=1.clear=0: 00:00:01.0750065
TestPerformance.set=Insert.threads=05.cache=1.clear=0: 00:00:00.6870007
TestPerformance.set=Insert.threads=10.cache=1.clear=0: 00:00:00.8658117
TestPerformance.set=Insert.threads=20.cache=1.clear=0: 00:00:00.8693524
TestPerformance.set=InsertObject.threads=01.cache=1.clear=0: 00:00:00.5815438
TestPerformance.set=InsertObject.threads=02.cache=1.clear=0: 00:00:00.5966730
TestPerformance.set=InsertObject.threads=05.cache=1.clear=0: 00:00:00.7877734
TestPerformance.set=InsertObject.threads=10.cache=1.clear=0: 00:00:00.9724163
TestPerformance.set=InsertObject.threads=20.cache=1.clear=0: 00:00:00.9999447
TestPerformance.set=InsertOrUpdate.threads=01.cache=1.clear=0: 00:00:03.0234816
TestPerformance.set=InsertOrUpdate.threads=02.cache=1.clear=0: 00:00:01.6098284
TestPerformance.set=InsertOrUpdate.threads=05.cache=1.clear=0: 00:00:01.1040860
TestPerformance.set=InsertOrUpdate.threads=10.cache=1.clear=0: 00:00:01.2332091
TestPerformance.set=InsertOrUpdate.threads=20.cache=1.clear=0: 00:00:01.2416994
TestPerformance.set=InsertOrUpdateObject.threads=01.cache=1.clear=0: 00:00:00.5850865
TestPerformance.set=InsertOrUpdateObject.threads=02.cache=1.clear=0: 00:00:00.6005924
TestPerformance.set=InsertOrUpdateObject.threads=05.cache=1.clear=0: 00:00:00.7837366
TestPerformance.set=InsertOrUpdateObject.threads=10.cache=1.clear=0: 00:00:00.9969544
TestPerformance.set=InsertOrUpdateObject.threads=20.cache=1.clear=0: 00:00:01.0329692
TestPerformance.set=InsertWithIdentity.threads=01.cache=1.clear=0: 00:00:02.2242455
TestPerformance.set=InsertWithIdentity.threads=02.cache=1.clear=0: 00:00:01.1841582
TestPerformance.set=InsertWithIdentity.threads=05.cache=1.clear=0: 00:00:00.7725962
TestPerformance.set=InsertWithIdentity.threads=10.cache=1.clear=0: 00:00:01.0034897
TestPerformance.set=InsertWithIdentity.threads=20.cache=1.clear=0: 00:00:00.9153585
TestPerformance.set=InsertWithIdentityObject.threads=01.cache=1.clear=0: 00:00:00.5725708
TestPerformance.set=InsertWithIdentityObject.threads=02.cache=1.clear=0: 00:00:00.5840798
TestPerformance.set=InsertWithIdentityObject.threads=05.cache=1.clear=0: 00:00:00.7523923
TestPerformance.set=InsertWithIdentityObject.threads=10.cache=1.clear=0: 00:00:00.9749190
TestPerformance.set=InsertWithIdentityObject.threads=20.cache=1.clear=0: 00:00:01.0524871
TestPerformance.set=MixedAll.threads=01.cache=1.clear=0: 00:00:01.3984295
TestPerformance.set=MixedAll.threads=02.cache=1.clear=0: 00:00:03.3527352
TestPerformance.set=MixedAll.threads=05.cache=1.clear=0: 00:00:05.3791372
TestPerformance.set=MixedAll.threads=10.cache=1.clear=0: 00:00:06.2213494
TestPerformance.set=MixedAll.threads=20.cache=1.clear=0: 00:00:06.9506050
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.0264842
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:05.7701840
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:09.3030392
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:11.7290079
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:11.1942493
TestPerformance.set=MixedObject.threads=01.cache=1.clear=0: 00:00:00.4930088
TestPerformance.set=MixedObject.threads=02.cache=1.clear=0: 00:00:00.3323332
TestPerformance.set=MixedObject.threads=05.cache=1.clear=0: 00:00:00.3383343
TestPerformance.set=MixedObject.threads=10.cache=1.clear=0: 00:00:00.3918670
TestPerformance.set=MixedObject.threads=20.cache=1.clear=0: 00:00:00.4359089
TestPerformance.set=Select.threads=01.cache=1.clear=0: 00:00:01.7857345
TestPerformance.set=Select.threads=02.cache=1.clear=0: 00:00:00.9263678
TestPerformance.set=Select.threads=05.cache=1.clear=0: 00:00:00.5275096
TestPerformance.set=Select.threads=10.cache=1.clear=0: 00:00:00.5120008
TestPerformance.set=Select.threads=20.cache=1.clear=0: 00:00:00.4819712
TestPerformance.set=Update.threads=01.cache=1.clear=0: 00:00:00.6942371
TestPerformance.set=Update.threads=02.cache=1.clear=0: 00:00:00.6987543
TestPerformance.set=Update.threads=05.cache=1.clear=0: 00:00:01.0705387
TestPerformance.set=Update.threads=10.cache=1.clear=0: 00:00:01.1300886
TestPerformance.set=Update.threads=20.cache=1.clear=0: 00:00:01.1475846
TestPerformance.set=UpdateObject.threads=01.cache=1.clear=0: 00:00:00.5982487
TestPerformance.set=UpdateObject.threads=02.cache=1.clear=0: 00:00:00.5765330
TestPerformance.set=UpdateObject.threads=05.cache=1.clear=0: 00:00:00.7642388
TestPerformance.set=UpdateObject.threads=10.cache=1.clear=0: 00:00:00.9889334
TestPerformance.set=UpdateObject.threads=20.cache=1.clear=0: 00:00:01.0329686
```

AFTER
```
TestPerformance.set=Delete.threads=01.cache=1.clear=0: 00:00:01.3688965
TestPerformance.set=Delete.threads=02.cache=1.clear=0: 00:00:00.7217037
TestPerformance.set=Delete.threads=05.cache=1.clear=0: 00:00:00.4013671
TestPerformance.set=Delete.threads=10.cache=1.clear=0: 00:00:00.3673564
TestPerformance.set=Delete.threads=20.cache=1.clear=0: 00:00:00.4429342
TestPerformance.set=DeleteObject.threads=01.cache=1.clear=0: 00:00:00.1481411
TestPerformance.set=DeleteObject.threads=02.cache=1.clear=0: 00:00:00.1142643
TestPerformance.set=DeleteObject.threads=05.cache=1.clear=0: 00:00:00.2092032
TestPerformance.set=DeleteObject.threads=10.cache=1.clear=0: 00:00:00.2552433
TestPerformance.set=DeleteObject.threads=20.cache=1.clear=0: 00:00:00.2342258
TestPerformance.set=Insert.threads=01.cache=1.clear=0: 00:00:02.0267384
TestPerformance.set=Insert.threads=02.cache=1.clear=0: 00:00:01.0400206
TestPerformance.set=Insert.threads=05.cache=1.clear=0: 00:00:00.7123472
TestPerformance.set=Insert.threads=10.cache=1.clear=0: 00:00:00.8648114
TestPerformance.set=Insert.threads=20.cache=1.clear=0: 00:00:00.9308803
TestPerformance.set=InsertObject.threads=01.cache=1.clear=0: 00:00:00.5835842
TestPerformance.set=InsertObject.threads=02.cache=1.clear=0: 00:00:00.5924541
TestPerformance.set=InsertObject.threads=05.cache=1.clear=0: 00:00:00.8608301
TestPerformance.set=InsertObject.threads=10.cache=1.clear=0: 00:00:01.0935305
TestPerformance.set=InsertObject.threads=20.cache=1.clear=0: 00:00:01.1235594
TestPerformance.set=InsertOrUpdate.threads=01.cache=1.clear=0: 00:00:03.0409258
TestPerformance.set=InsertOrUpdate.threads=02.cache=1.clear=0: 00:00:01.6075777
TestPerformance.set=InsertOrUpdate.threads=05.cache=1.clear=0: 00:00:01.0841962
TestPerformance.set=InsertOrUpdate.threads=10.cache=1.clear=0: 00:00:01.2192236
TestPerformance.set=InsertOrUpdate.threads=20.cache=1.clear=0: 00:00:01.2932236
TestPerformance.set=InsertOrUpdateObject.threads=01.cache=1.clear=0: 00:00:00.5795714
TestPerformance.set=InsertOrUpdateObject.threads=02.cache=1.clear=0: 00:00:00.6156022
TestPerformance.set=InsertOrUpdateObject.threads=05.cache=1.clear=0: 00:00:00.9063504
TestPerformance.set=InsertOrUpdateObject.threads=10.cache=1.clear=0: 00:00:01.1225530
TestPerformance.set=InsertOrUpdateObject.threads=20.cache=1.clear=0: 00:00:01.0925240
TestPerformance.set=InsertWithIdentity.threads=01.cache=1.clear=0: 00:00:02.0598161
TestPerformance.set=InsertWithIdentity.threads=02.cache=1.clear=0: 00:00:01.1392014
TestPerformance.set=InsertWithIdentity.threads=05.cache=1.clear=0: 00:00:00.7167277
TestPerformance.set=InsertWithIdentity.threads=10.cache=1.clear=0: 00:00:00.9493908
TestPerformance.set=InsertWithIdentity.threads=20.cache=1.clear=0: 00:00:00.9543955
TestPerformance.set=InsertWithIdentityObject.threads=01.cache=1.clear=0: 00:00:00.6069756
TestPerformance.set=InsertWithIdentityObject.threads=02.cache=1.clear=0: 00:00:00.6094809
TestPerformance.set=InsertWithIdentityObject.threads=05.cache=1.clear=0: 00:00:00.8968710
TestPerformance.set=InsertWithIdentityObject.threads=10.cache=1.clear=0: 00:00:01.0599949
TestPerformance.set=InsertWithIdentityObject.threads=20.cache=1.clear=0: 00:00:01.1110429
TestPerformance.set=MixedAll.threads=01.cache=1.clear=0: 00:00:01.3943687
TestPerformance.set=MixedAll.threads=02.cache=1.clear=0: 00:00:02.0160013
TestPerformance.set=MixedAll.threads=05.cache=1.clear=0: 00:00:02.1891704
TestPerformance.set=MixedAll.threads=10.cache=1.clear=0: 00:00:02.5802990
TestPerformance.set=MixedAll.threads=20.cache=1.clear=0: 00:00:03.3414328
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.0330294
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:03.6962151
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:03.7004956
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:04.3561462
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:05.4310002
TestPerformance.set=MixedObject.threads=01.cache=1.clear=0: 00:00:00.5195115
TestPerformance.set=MixedObject.threads=02.cache=1.clear=0: 00:00:00.3443415
TestPerformance.set=MixedObject.threads=05.cache=1.clear=0: 00:00:00.3613519
TestPerformance.set=MixedObject.threads=10.cache=1.clear=0: 00:00:00.4275196
TestPerformance.set=MixedObject.threads=20.cache=1.clear=0: 00:00:00.4304133
TestPerformance.set=Select.threads=01.cache=1.clear=0: 00:00:02.0895978
TestPerformance.set=Select.threads=02.cache=1.clear=0: 00:00:00.9958491
TestPerformance.set=Select.threads=05.cache=1.clear=0: 00:00:00.5515831
TestPerformance.set=Select.threads=10.cache=1.clear=0: 00:00:00.5885791
TestPerformance.set=Select.threads=20.cache=1.clear=0: 00:00:00.5575479
TestPerformance.set=Update.threads=01.cache=1.clear=0: 00:00:00.7822896
TestPerformance.set=Update.threads=02.cache=1.clear=0: 00:00:00.7282045
TestPerformance.set=Update.threads=05.cache=1.clear=0: 00:00:00.9629302
TestPerformance.set=Update.threads=10.cache=1.clear=0: 00:00:01.2356669
TestPerformance.set=Update.threads=20.cache=1.clear=0: 00:00:01.2701911
TestPerformance.set=UpdateObject.threads=01.cache=1.clear=0: 00:00:00.5910557
TestPerformance.set=UpdateObject.threads=02.cache=1.clear=0: 00:00:00.5997031
TestPerformance.set=UpdateObject.threads=05.cache=1.clear=0: 00:00:00.8828501
TestPerformance.set=UpdateObject.threads=10.cache=1.clear=0: 00:00:01.1015409
TestPerformance.set=UpdateObject.threads=20.cache=1.clear=0: 00:00:01.1310608
```

for non-linq queries and test runs with single query there is no changes in performance (it is already high enough), but for test runs with several different linq qieries fighting for the same cache (`MixedLinq` and `MixedAll`) significant performance boost could be observed for multithreaded environments:

```
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.0264842
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:05.7701840
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:09.3030392
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:11.7290079
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:11.1942493
```
VS
```
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.0330294
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:03.6962151
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:03.7004956
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:04.3561462
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:05.4310002
```

unfortunatelly, extra threads still add their overhead instead of performance boost, but take into account that those are runs without real database access.

MaceWindu commented:
Another round of improvements:

```
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.0330294
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:03.6962151
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:03.7004956
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:04.3561462
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:05.4310002
```
VS
```
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:01.9875546
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:03.4790465
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:03.7506074
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:04.2879782
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:05.1334197
```

MaceWindu commented:
Looks like current cache structure not sutable for multithreaded access. Redesign results (at some allocations cost):

```
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.2689869
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:01.2261460
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:00.7276843
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:00.7997750
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:00.7772296
```

MaceWindu commented:
Mono crash in travis looks like this issue https://github.com/nunit/nunit-console/issues/211

MaceWindu commented:
Added extra tests and fixed bugs. Performance drop observed on 20 threads, but still high
```
TestPerformance.set=MixedLinq.threads=01.cache=1.clear=0: 00:00:02.2919112
TestPerformance.set=MixedLinq.threads=02.cache=1.clear=0: 00:00:01.1634750
TestPerformance.set=MixedLinq.threads=05.cache=1.clear=0: 00:00:00.7000284
TestPerformance.set=MixedLinq.threads=10.cache=1.clear=0: 00:00:00.7982900
TestPerformance.set=MixedLinq.threads=20.cache=1.clear=0: 00:00:01.4161880
```

MaceWindu commented:
Mono crashed again. Looks like it support multithreading only if number of threads = 1 ...

sdanyliv commented on Source/Linq/Query.cs:
```diff
@@ -128,94 +127,139 @@ IEnumerable<T> MakeEnumerable(QueryContext qc, IDataContext dc, Expression expr,
 		#region GetInfo
 
-		static          Query<T> _first;
+		static          List<Query<T>> _orderedCache = new List<Query<T>>(CacheSize);
```
Maybe LinkedList is more effective here?

MaceWindu replied:
Will do performance/memory tests during weekend

MaceWindu replied:
Looks like List performs better, see bechmarks report below

igor-tkachev commented on Source/Linq/Query.cs:
```diff
@@ -128,94 +127,139 @@ IEnumerable<T> MakeEnumerable(QueryContext qc, IDataContext dc, Expression expr,
+							var first = _orderedCache[0];
+							_orderedCache[0] = query;
+							_orderedCache[oldIndex] = first;
+						}
```
Previous implementation moved the most frequently used queries to the beginning of the list. This one will swap two queries that may lead to scan the whole list every time if two queries are first and last.

Ideally, we need some ordering here by usage counter.

MaceWindu replied:
My bad, wasn't my intention - will fix it.
I don't think simple ordering by counter will be a good idea, as it could lead to cache poisoning by rarely but intensively used queries.

MaceWindu replied:
Fixed

MaceWindu commented:
- Fixed quiery reordering
 - Added benchmark for List and LinkedList (not added to any project as benchmarks project not merged yet, see #739 )

Looks like LinkedList behaves better on small cache size (10) but on out normal cache size(100) it is slower than List. Memory allocations look the same for both collections.

``` ini

BenchmarkDotNet=v0.10.9, OS=Windows 10 Redstone 1 (10.0.14393)
Processor=Intel Core i7-3770K CPU 3.50GHz (Ivy Bridge), ProcessorCount=8
Frequency=3417997 Hz, Resolution=292.5690 ns, Timer=TSC
  [Host]     : .NET Framework 4.7 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.2101.1
  DefaultJob : .NET Framework 4.7 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.2101.1


```
 |    Method | Threads | CacheSize | FillFactor | Collection |      Mean |     Error |    StdDev |     Gen 0 |  Allocated |
 |---------- |-------- |---------- |----------- |----------- |----------:|----------:|----------:|----------:|-----------:|
 | **Benchmark** |       **1** |        **10** |        **0,5** | **LinkedList** |  **2.031 ms** | **0.0049 ms** | **0.0045 ms** |   **78.1250** |  **328.18 KB** |
 | **Benchmark** |       **1** |        **10** |        **0,5** |       **List** |  **2.402 ms** | **0.0121 ms** | **0.0113 ms** |   **78.1250** |  **328.12 KB** |
 | **Benchmark** |       **1** |        **10** |        **1,5** | **LinkedList** |  **2.976 ms** | **0.0125 ms** | **0.0104 ms** |  **144.5313** |  **603.37 KB** |
 | **Benchmark** |       **1** |        **10** |        **1,5** |       **List** |  **3.184 ms** | **0.0548 ms** | **0.0513 ms** |  **125.0000** |   **524.8 KB** |
 | **Benchmark** |       **1** |       **100** |        **0,5** | **LinkedList** |  **9.472 ms** | **0.0259 ms** | **0.0230 ms** |  **500.0000** | **2080.84 KB** |
 | **Benchmark** |       **1** |       **100** |        **0,5** |       **List** |  **7.449 ms** | **0.0248 ms** | **0.0232 ms** |  **503.1250** | **2080.12 KB** |
 | **Benchmark** |       **1** |       **100** |        **1,5** | **LinkedList** | **16.547 ms** | **0.0596 ms** | **0.0497 ms** |  **968.7500** | **4096.03 KB** |
 | **Benchmark** |       **1** |       **100** |        **1,5** |       **List** | **12.418 ms** | **0.0293 ms** | **0.0274 ms** |  **968.7500** | **4016.49 KB** |
 | **Benchmark** |       **2** |        **10** |        **0,5** | **LinkedList** |  **2.338 ms** | **0.0146 ms** | **0.0137 ms** |   **78.1250** |  **332.19 KB** |
 | **Benchmark** |       **2** |        **10** |        **0,5** |       **List** |  **2.622 ms** | **0.0154 ms** | **0.0144 ms** |   **78.1250** |  **332.17 KB** |
 | **Benchmark** |       **2** |        **10** |        **1,5** | **LinkedList** |  **3.241 ms** | **0.0144 ms** | **0.0135 ms** |  **147.3958** |  **608.28 KB** |
 | **Benchmark** |       **2** |        **10** |        **1,5** |       **List** |  **3.458 ms** | **0.0091 ms** | **0.0085 ms** |  **128.9063** |  **531.15 KB** |
 | **Benchmark** |       **2** |       **100** |        **0,5** | **LinkedList** |  **9.886 ms** | **0.0488 ms** | **0.0456 ms** |  **500.0000** | **2088.98 KB** |
 | **Benchmark** |       **2** |       **100** |        **0,5** |       **List** |  **7.846 ms** | **0.0569 ms** | **0.0532 ms** |  **500.0000** | **2087.87 KB** |
 | **Benchmark** |       **2** |       **100** |        **1,5** | **LinkedList** | **17.144 ms** | **0.1572 ms** | **0.1393 ms** | **1000.0000** | **4125.45 KB** |
 | **Benchmark** |       **2** |       **100** |        **1,5** |       **List** | **12.811 ms** | **0.0605 ms** | **0.0566 ms** |  **984.3750** | **4083.66 KB** |
 | **Benchmark** |       **5** |        **10** |        **0,5** | **LinkedList** |  **3.222 ms** | **0.0366 ms** | **0.0343 ms** |   **82.0313** |  **344.52 KB** |
 | **Benchmark** |       **5** |        **10** |        **0,5** |       **List** |  **3.705 ms** | **0.0572 ms** | **0.0535 ms** |   **82.0313** |  **344.42 KB** |
 | **Benchmark** |       **5** |        **10** |        **1,5** | **LinkedList** |  **4.333 ms** | **0.0388 ms** | **0.0363 ms** |  **148.4375** |  **620.35 KB** |
 | **Benchmark** |       **5** |        **10** |        **1,5** |       **List** |  **4.606 ms** | **0.0642 ms** | **0.0569 ms** |  **125.0000** |  **543.19 KB** |
 | **Benchmark** |       **5** |       **100** |        **0,5** | **LinkedList** | **11.761 ms** | **0.1699 ms** | **0.1506 ms** |  **500.0000** | **2105.21 KB** |
 | **Benchmark** |       **5** |       **100** |        **0,5** |       **List** |  **9.587 ms** | **0.1848 ms** | **0.1815 ms** |  **500.0000** | **2105.43 KB** |
 | **Benchmark** |       **5** |       **100** |        **1,5** | **LinkedList** | **20.238 ms** | **0.1441 ms** | **0.1125 ms** | **1000.0000** | **4199.14 KB** |
 | **Benchmark** |       **5** |       **100** |        **1,5** |       **List** | **15.005 ms** | **0.1396 ms** | **0.1238 ms** | **1000.0000** | **4195.33 KB** |
 | **Benchmark** |      **10** |        **10** |        **0,5** | **LinkedList** |  **3.999 ms** | **0.0423 ms** | **0.0396 ms** |   **82.2917** |  **351.58 KB** |
 | **Benchmark** |      **10** |        **10** |        **0,5** |       **List** |  **4.762 ms** | **0.0548 ms** | **0.0512 ms** |   **81.2500** |  **352.07 KB** |
 | **Benchmark** |      **10** |        **10** |        **1,5** | **LinkedList** |  **5.321 ms** | **0.0370 ms** | **0.0328 ms** |  **148.4375** |  **630.08 KB** |
 | **Benchmark** |      **10** |        **10** |        **1,5** |       **List** |  **5.713 ms** | **0.0558 ms** | **0.0495 ms** |  **132.8125** |  **551.06 KB** |
 | **Benchmark** |      **10** |       **100** |        **0,5** | **LinkedList** | **14.499 ms** | **0.1032 ms** | **0.0965 ms** |  **515.6250** | **2120.19 KB** |
 | **Benchmark** |      **10** |       **100** |        **0,5** |       **List** | **11.913 ms** | **0.1597 ms** | **0.1493 ms** |  **515.6250** | **2122.25 KB** |
 | **Benchmark** |      **10** |       **100** |        **1,5** | **LinkedList** | **24.986 ms** | **0.2263 ms** | **0.2006 ms** | **1031.2500** | **4256.54 KB** |
 | **Benchmark** |      **10** |       **100** |        **1,5** |       **List** | **18.267 ms** | **0.1859 ms** | **0.1739 ms** | **1031.2500** | **4230.08 KB** |
 | **Benchmark** |      **20** |        **10** |        **0,5** | **LinkedList** |  **4.100 ms** | **0.0336 ms** | **0.0314 ms** |   **85.9375** |  **354.63 KB** |
 | **Benchmark** |      **20** |        **10** |        **0,5** |       **List** |  **4.831 ms** | **0.0679 ms** | **0.0635 ms** |   **85.9375** |  **354.58 KB** |
 | **Benchmark** |      **20** |        **10** |        **1,5** | **LinkedList** |  **5.508 ms** | **0.0457 ms** | **0.0428 ms** |  **148.9583** |  **636.16 KB** |
 | **Benchmark** |      **20** |        **10** |        **1,5** |       **List** |  **6.672 ms** | **0.1330 ms** | **0.2531 ms** |  **132.8125** |   **558.1 KB** |
 | **Benchmark** |      **20** |       **100** |        **0,5** | **LinkedList** | **15.008 ms** | **0.1931 ms** | **0.1806 ms** |  **515.6250** | **2127.19 KB** |
 | **Benchmark** |      **20** |       **100** |        **0,5** |       **List** | **12.359 ms** | **0.1292 ms** | **0.1079 ms** |  **515.6250** | **2130.76 KB** |
 | **Benchmark** |      **20** |       **100** |        **1,5** | **LinkedList** | **26.171 ms** | **0.1805 ms** | **0.1688 ms** | **1031.2500** | **4254.59 KB** |
 | **Benchmark** |      **20** |       **100** |        **1,5** |       **List** | **19.093 ms** | **0.1880 ms** | **0.1759 ms** | **1016.6667** | **4224.34 KB** |

MaceWindu commented:
Merged with latest changes from master
